[wifi] Event based connect/reconnect

TD-er opened the pull request:
Now using the core WiFiEvents to act on:

- connect
- disconnect
- gotIP

This will make polling for a status no longer needed.
It also allows to get a better understanding in why diconnects happen, since the disconnect reason is given in the events.
This extra information is also present on the information panel on the webinterface and also in the JSON output.

The entire connect/disconnect/reconnect is now done completely asynchronous and thus takes much less resources from the ESP.
A disconnect will now also send an event: "WiFi#Disconnected"

For deepsleep this means there has to be a minimum awake time set and perhaps use rules to set the ESP into sleep again when done.
Typical connection times to an accesspoint are about 3 seconds + about 1 second for DHCP.

This also resolves the issues where services my try to connect while there is a WiFi connection, but not yet an IP address set.

I already noticed a lot of NTP servers from pool.ntp.org are non responsive the last few weeks, so that may still take some time, but that's another issue.

TD-er committed:
[wifi] Event based connect/reconnect

Now using the core WiFiEvents to act on:

- connect
- disconnect
- gotIP

This will make polling for a status no longer needed.
It also allows to get a better understanding in why diconnects happen, since the disconnect reason is given in the events.
This extra information is also present on the information panel on the webinterface and also in the JSON output.

The entire connect/disconnect/reconnect is now done completely asynchronous and thus takes much less resources from the ESP.
A disconnect will now also send an event: "WiFi#Disconnected"

For deepsleep this means there has to be a minimum awake time set and perhaps use rules to set the ESP into sleep again when done.
Typical connection times to an accesspoint are about 3 seconds + about 1 second for DHCP.

This also resolves the issues where services my try to connect while there is a WiFi connection, but not yet an IP address set.

I already noticed a lot of NTP servers from pool.ntp.org are non responsive the last few weeks, so that may still take some time, but that's another issue.

TD-er commented:
Tested it extensively, but please let others test before we merge this one.

Oxyandy commented:
So how to build and test ?

Oxyandy commented:
I have Git Hub desktop, if I update source & compile is that enough, then use Dev firmware ?
Think I get it, do this - open in Atom compile
![charthhhhhh](https://user-images.githubusercontent.com/36609061/37859150-fb67f3f4-2f63-11e8-9fd4-cc80622df11b.jpg)

TD-er commented:
I will build a test version later this morning

TD-er commented:
[here](https://www.dropbox.com/s/lss6zh5i2mqpkhu/ESPeasy_event_driven_WiFi.rar?dl=0) a test build of this PR.

micropet commented:
Starts normal. Otherwise, I do not notice a difference.

TD-er commented:
Working is a very nice feature given last few days :)
Please also have a look at the info page, console log and json page

micropet commented:
![json](https://user-images.githubusercontent.com/2838584/37862375-fd3e9678-2f4b-11e8-812a-5ea378f79f72.png)

![main](https://user-images.githubusercontent.com/2838584/37862390-3ba342ec-2f4c-11e8-9edc-938a7c2b353a.png)

![log](https://user-images.githubusercontent.com/2838584/37862408-9104b892-2f4c-11e8-99ea-76af240e73de.png)

I have so many devices on the network now,
that I'm not sure if I've flashed the right one.
This one reacts very slowly in the web interface.

TD-er commented:
That's the new version.
I meant tools -> info page.
But json told me enough.

I will test the speed here.
You can also see how long WiFi keeps being connected.

micropet commented:
![sysinfo](https://user-images.githubusercontent.com/2838584/37862424-edc81fce-2f4c-11e8-8a26-97221454df2a.png)

Grovkillen commented:
Wow @TD-er this is awesome work you have done!

TD-er commented:
@micropet I was just looking at your screenshot and the CPU load is quite high.
Do you have a lot of rules?
There's another PR for that ;)
Later today, I will try to find some time to merge both in a test build.

I've tested here on my Chromebook (that one is running on a Celeron and no virusscanner) and the pages appear in about 300 - 400 msec. That's not what I consider slow.
On my i7 laptop pageload may take up to 1.5 seconds, but that's due to the virusscanner :(

micropet commented:
I only have the visual comparison to the other devices.
For the others, the pages build up much faster.
This Wemos has also no hardware connected.

Thats the Rules:

// ----- ESP-218 Dust only

On System#Boot do    //When the ESP boots, do
	Publish %sysname%/IP,%ip%
	timerSet,1,60      //Set Timer 1 for the next event in 60 seconds
	// timerSet,2,20
endon

On Rules#Timer=1 do  //When Timer1 expires, do
	Publish %sysname%/IP,%ip%
	Publish %sysname%/Time,%systime%
	Publish %sysname%/Uptime,%uptime%
        Publish %sysname%/RSSI,%rssi%
	timerSet,1,60       //Resets the Timer 1 for another 60 seconds
endon
// -----

susisstrolch commented:
Just tried your precompiled version on WeMOS Mini, using the dev_ESP8266_4096_core2_4_0.
With this build NTP doesn't work anymore.
I'm using my Synology NTP server, which is fine for all other inhouse devices.

micropet commented:
How do you see how fast a page builds up?

micropet commented:
i have test_ESP8266_4096 from today morning. NTP works.

susisstrolch commented:
also the core_2_4_0 version?

micropet commented:
I do not know that. Is from the file of the Dropbox of Gijs

TD-er commented:
About NTP, I've seen some issues also, that it may take several attempts (retry is a minute) to get it.
But that was already an issue before this rewrite and is rather high on my 'to investigate' list.
It's good to know it also borks on local NTP servers.
I tried from pool.ntp.org and that returns another IP-address every time you do a new request. But quite a few of these do not react. I've tried a few, even from my laptop.

@micropet You said you had no hardware connected. But do you have devices defined in ESPeasy?
I've seen sometimes these plugins try to do stuff, which will timeout. But that busy waiting may cause some delays.

Also a lot of services now check for presence of the host they try to connect, for which I use ping.
But I've noticed this ping may also cause some delay, especially when the host is offline.
I will also look into that an try to resolve that in an asynchronous way.

In short; it is under investigation.

Did you also test whether it would connect for "the first time" and with deepsleep? (you may need to set the minimum awake time to about 5 or 6 seconds)

micropet commented:
![device_page_1](https://user-images.githubusercontent.com/2838584/37863644-95dbf280-2f61-11e8-9d8f-14b2d3e04888.png)

I have not configured any devices.
I will test Deepsleep later.

vestus commented on src/ESPEasy.ino:
```diff
@@ -307,9 +298,14 @@ void loop()
+    WiFiConnectRelaxed();
     wifiSetupConnect = false;
   }
+  if (wifiStatus != ESPEASY_WIFI_SERVICES_INITIALIZED) {
```
These may be better suited for an enum. 

Switch statement would also be more appropriate and accommodate future logic changes.

Oxyandy commented:
I attempt to make any changes, this is what I see..
Can anyone confirm please ?
However it seems, the changes do save on clicking submit. Ta
![broken](https://user-images.githubusercontent.com/36609061/37870715-f8570032-3027-11e8-9700-90fc3505e01c.jpg)

TD-er commented:
Do these errors happen all the time you're trying to save anything? Also after power cycle?
Can you see in the logs what the actual error is?

This is not related to the changes of this PR, but still something to look at.
It may be a hardware failure, but that is hard to tell for sure without logs (if anything is logged about the cause)

Oxyandy commented:
all the time at the moment, any time I try
```
14141875 : WD : Uptime 231 ConnectFailures 0 FreeMem 13248
14148570 : FS : Error while reading/writing config.dat in 1879
14148582 : FS : Error while reading/writing config.dat in 1879
14150043 : SYS : -60.00
14150045 : EVENT: WiFiSig4#WiFiSig4=-60.00
```
Before powering down, checked & made other changes to other devices
I edited Device Systime & then a Switch & yes same message about being not being able to save.
Right now GPIO14 is stuck, I can not change it, it doesn't match pinstate page either..
A command does not clear it either..
Powered off / on
It behaves normally, maybe I try again once GPIO14 (the only one I am using really except relay GPIO12)
locks up again, this has happened on many modules, I do not think I have ever tried to edit anything before while it has been 'stuck' - will check next it locks up, then if same, will try a different module

Oxyandy commented:
Yes, I can edit & save (Submit) anything until (as a symptom) GPIO14 gets stuck as '1'
(edit for clarity this is a whole new power cycle)
I got exactly the same message "FS : Error while reading/writing config.dat in 1879" with the 'Red Bar'
GPIO14 getting stuck really does happen fast on this firmware, I was using an older but still recent fimware release earlier today & it gets stuck far less often..
I guess forget rules as a bug-fix but rather look into pinstates vs the Devices page ??

Oxyandy commented:
Do you want me to take a brand new module flash with latest firmware. looking for the same symptoms:
1. GPIO14 shows as wrong '1' while it should be '0' - then - once this glitch has happened.
2. Try edit 'anything' on Devices

Grovkillen commented:
I can confirm the gpio 14 problem on one of my units. Currently I solve this using a rule with TaskValueSet to make sure the relay state (which I monitor using a switch device) is the correct one.

Oxyandy commented:
Ok I have been doing a bit of reading, but unsure
My GPIO14 is device 1, 
it gets stuck '1', I want to force it '0' - GPIO14,0 is not working in rules..
My Dummy is device 7
But then I get a bit confused - a bit of help will be awesome thanks
TaskValueSet ?,?,?

Grovkillen commented:
This is how I do it (my Pump#State is on task/device number 8):

```
on MQTT#Connected do
 Publish %sysname%/IP,%ip%
 timerSet,1,10
endon

on Rules#Timer=1 do
  if [Floater_Top#Floating]=1
   event,TurnOnPump=600
  else
   event,TurnOffPump
 endif
endon

on TurnOnPump do
   timerSet,1,%eventvalue% //no need to do the check first 10 minutes
   gpio,14,1
   TaskValueSet,8,1,1
   Publish %sysname%/Pump,Started
   Publish %sysname%/ExtraInfo,Pump started at %systime%
endon

on TurnOffPump do
 if [Pump#State]=1 and [Flow_Sensor#mL]=0. 00
   gpio,14,0
   TaskValueSet,8,1,0
   Publish %sysname%/Pump,Stopped
   Publish %sysname%/ExtraInfo,Pump stopped at %systime%
   timerSet,1,10
  else
   timerSet,1,10
  endIf
endon
```

Oxyandy commented:
@Grovkillen 
![back](https://user-images.githubusercontent.com/36609061/37873808-70a5f0d8-306f-11e8-83e2-b1e133e8ee64.jpg)
As above in the picture says it's fine but it doesnt change the result on device page.
This will be the last I talk about this on here, happy to take it over to the forum, but you are absolutely right, the ONLY thing which changes my switch from '1' to '0' on the device page is TaskValueSet 1,1,0 
- just one tiny problem, the switch no longer responds to any change, 
moving it up /down doesn't stop the pump.
Once this situation is stuck, it's always stuck... 
This rule actually stop reporting / working 

```
on fulltray4#switch do //my switch on GPIO14
if [fulltray4#switch]=1
gpio,12,0
publish domoticz/in,{"idx":26,"nvalue":0,"svalue":"floatswitch1%systime%"}
else
gpio,12,0
publish domoticz/in,{"idx":26,"nvalue":0,"svalue":"floatswitch0%systime%"}
endif
endon
```
Wonder if I moved switch to GPIO3 if it would change anything, never tried that.
Thanks for your input.

susisstrolch commented:
@TD-er just an Info about the test...core_2_4_0 - it simply leaves too less RAM (~4k) for propper testing. Maybe that's the reason for some of my issues.

TD-er commented:
Wasn't GPIO3 not connected to the hardware serial?

I just found why the NTP was so much slower with this patch.
Previously it would only start calling "now" when the WiFi was already setup.
And in calling `now()`, a call to an NTP server is being made. When this fails, retry again in a minute.
But now the WiFi initialization is asynchronous, so the first NTP query was already failing and thus the retry is already set to a minute.

This is fixed now and also the pool.ntp.org will pick a random one from 0...3 (.pool.ntp.org).
The DNS queries from this pool change about every minute, so if you want to retry because one of the servers is offline, you should try one from the other N.pool.ntp.org or else you have to wait for another minute.

@susisstrolch What version do you use (normal/dev/test) and how many plugins?

TD-er committed:
[NTP] Speedup NTP again with async WiFi

Previously it would only start calling "now" when the WiFi was already setup.
And in calling `now()`, a call to an NTP server is being made. When this fails, retry again in a minute.
But now the WiFi initialization is asynchronous, so the first NTP query was already failing and thus the retry is already set to a minute.

This is fixed now and also the pool.ntp.org will pick a random one from 0...3 (.pool.ntp.org).
The DNS queries from this pool change about every minute, so if you want to retry because one of the servers is offline, you should try one from the other `N.pool.ntp.org` or else you have to wait for another minute.

susisstrolch commented:
@TD-er I just used the test from your RAR.

TD-er commented:
On the info page you can see how many plugins are included.
The current max, with test set as #define is about 72?

Edit:
I just flashed the 2.4.0 core version and indeed, it is very limited on memory:

> Free Mem | 6960 (2216 - duringDataTX)

Oxyandy commented:
@TD-er "Wasn't GPIO3 not connected to the hardware serial?"
Yes correct on a Sonoff Basic it is, if I turn off serial logging, I had no problem with using GPIO3
I also have the option to try GPIO4 or 5 too.. 
simple soldering a wire from chip to unused radio module holes with a gang of header pins

TD-er commented:
I created a new build, including NTP fix and another normal_esp8266_core_2.4.0 build, which uses less memory.
[ESPeasy_event_driven_WiFi_NTP-fix](https://www.dropbox.com/s/62huui3fvkfdquv/ESPeasy_event_driven_WiFi_NTP-fix.rar?dl=0)

susisstrolch commented:
@TD-er: looks much better now
`

Unit | 11
-- | --
Local Time | 2018-03-25 13:57:34
Uptime | 0 days 0 hours 2 minutes
Load | 5% (LC=19010)
Free Mem | 11736 (9424 - sendContentBlocking)

`
